### PR TITLE
Solver dd

### PR DESCRIPTION
**Pull request by jobordner**
Flux-correction

   Added flux data objects Face, FaceFluxes, and FluxData to store and
   manipulate fluxes, and added MethodFluxCorrect Method to perform
   flux corrections.  Current status is incomplete--flux data refresh
   requires debugging, and flux correction in MethodFluxCorrect is
   incomplete.

New Refresh

   Refresh has been overhauled in to include buffering.  This ensures
   that ghost zone data are not copied into a block's fields before
   the block is ready to receive them.  New refresh code is in
   control_new_refresh.cpp.  Note old refresh code is still in
   control_refresh.cpp  Robustness is significantly improved.

Solver DD

   The "DD" solver (domain-decomposition) applies BiCG-STAB to
   root-level tree solves independently (and in parallel), using
   boundary conditions from an MG0 root-level solve.  An optional
   final Jacobi smoother can be applied to the solution to smooth out
   the solution along domain boundaries.  While the solver
   implementation is complete and has been tested up through N0=2048^3
   root-grid size with 4 levels of refinement on 114K cores, the
   solution quality should be (and will be) evaluated for inaccuracies
   on root-grid level boundaries due to domain decomposition.

Testing

   Additional tests for cosmology and collapse tests (both
   particle-only and gas only) have been added, and include Mg0, CG,
   BiCG-STAB, DD, and HG (Mg0-preconditioned BiCG-STAB) solvers.  Note
   that collapse tests require double precision due to using cgs
   units.  (Regression tests still "pass" for single-precision,
   collapse tests are just not run).

Summary of other significant changes

  - Build: added "smp" variable to compile and test Charm++ smp builds
  - Build: updated "diff" make target to include staged and unstaged files
  - Cello: added cello_float
  - Cello: added more cello namespace accessor methods (::num_children() etc.)
  - Charm: changed mesh.ci and main*.ci to *.CI to include cpp macros
  - Doc: added design-flux document
  - Doc: updated 'getting-started' with corrections from Joshua Smith
  - Initial: updated collapse IC's
  - Initial: updated random number generators in Initial[Pm|SedovRandom|Soup]
  - IO: added throttling and staggering input for MUSIC IC's
  - Mesh: support mesh levels < 0, e.g. multigrid coarse grids
  - Mesh: updated block-process mapping to improve locality when nb > np
  - Perf: added load-balancing performance metrics
  - Perf: added schedule for Projections monitoring
  - Refresh: overhauled refresh to add buffering: control_new_refresh.cpp
  - Solver: added missing refresh before BiCG-STAB preconditioner
  - Test: added test_[cosmo|collapse]* input files for regression tests
  - Test: added unit tests for Sync, Face, FaceFluxes, FluxData
  - Test: modified some regression tests (e.g. particle) to not run so long
  - Test: moved unit tests outside of loops to reduce test output size

Numerous bug fixes have been made, include a bug in particle handling
(seg fault could occur when adding particles to an empty batch), and
several memory leaks.

### CODE REVIEW

File: src/Cello/simulation_Simulation.cpp [hunk 1]
```diff
@@ -948,24 +969,31 @@ void Simulation::r_monitor_performance(CkReductionMsg * msg)
...
+  for (int i=hierarchy_->min_level(); i<=hierarchy_->max_level(); i++) {
+    const long long num_blocks_level = counters_reduce[m++]; // NL
+    monitor()->print("performance","simulation num-blocks-level %d %ld",
 		     i,num_blocks_level);
     num_total_blocks += num_blocks_level;
-    if (i>0)
-      num_leaf_blocks += (num_blocks_level - num_blocks_level/num_child_blocks);
+    // compute leaf blocks given number of blocks per level
+    if (i>0) num_leaf_blocks +=
+	(num_blocks_level - num_blocks_level/cello::num_children());
```
Comment (bwoshea) [inline src/Cello/simulation_Simulation.cpp]: if something like num_blocks_level is an integer and cello::num_children() is also an integer, will this produce the desired behavior?

File: src/Cello/simulation_Simulation.cpp [hunk 2]
```diff
@@ -985,9 +1013,30 @@ void Simulation::r_monitor_performance(CkReductionMsg * msg)
...
-	  m,n, (m == n) );
+  const long long max_proc_blocks    = counters_reduce[m++]; // 10
+  const long long max_proc_particles = counters_reduce[m++]; // 11
+  
+  monitor()->print
+    ("Performance","simulation max-proc-blocks %d",  max_proc_blocks);
+  monitor()->print
+    ("Performance","simulation max-proc-particles %d", max_proc_particles);
+
+  const double avg_proc_blocks = 1.0*num_blocks_total/CkNumPes();
```
Comment (bwoshea) [inline src/Cello/simulation_Simulation.cpp]: As mentioned above, you're mixing precision on the right hand side.  Is that going to work out OK for all compilers?

Comment (bwoshea) [reply src/Cello/simulation_Simulation.cpp]: Answering my own question:  it seems that when you mix float/double and int, it always gets promoted to float/double.  So, the 1.0 ensures that the compiler will Do The Right Thing on the RHS.

File: src/Cello/simulation_Simulation.cpp (hunk 1 above)
Comment (jobordner) [reply src/Cello/simulation_Simulation.cpp]: @bwoshea In general no, but here yes: since all blocks refine fully into num_children() sub-blocks, the number of blocks in a refined level (i > 0) will always be evenly divisible by num_children().

File: SConstruct [hunk 3]
```diff
@@ -10,16 +10,16 @@ import socket
...
 
-new_output = 0
+new_refresh = 1
 
 #----------------------------------------------------------------------
-# Temporary setting for using new PPM routines from enzo-dev
+# Temporary setting for using new Output implementation
 #----------------------------------------------------------------------
 
-new_ppm = 1
```
Comment (aemerick) [inline SConstruct]: Why is this parameter getting removed? Or are the old PPM routines getting stripped, making this unnecessary?

Comment (jobordner) [reply SConstruct]: @aemerick I think I did a quick grep on *pp files and nothing came up, so I assumed this flag was obsolete; but I see that's because NEW_PPM is confined to *.F files.

@anyone Still, is there any reason to keep the old ppm_de.F, [xyz]euler_sweep.F code in Enzo-E?

Comment (gregbryan) [reply SConstruct]: If it is unused, I see no objection to removing it.

Comment (aemerick) [reply SConstruct]: Ah ok I see. I vote stripping out all of the old PPM code and leave just the “new”/enzo-dev PPM code, getting rid of this flag in the process. I don’t see any reason to keep around the old versions. 

Happy to submit this PR.

File: src/Cello/control_new_refresh.cpp [hunk 4]
```diff
@@ -0,0 +1,816 @@
...
+      // ...extract particle position arrays
+
+      double xa[np],ya[np],za[np];
+      particle.position(it,ib,xa,ya,za);
+
+      // ...initialize mask used for scatter and delete
+      // ...and corresponding particle indices
+
+      bool mask[np];
+      int index[np];
```
Comment (aemerick) [inline src/Cello/control_new_refresh.cpp:772]: I was looking through this to implement the local particle copy thing and noticed this. Wanted to flag it here to remember. But should this be `bool * mask = new bool [np];` and likewise for `index` (with an associated `delete` later on)?

Comment (jobordner) [reply src/Cello/control_new_refresh.cpp:772]: You're correct, this should be updated.  Thanks!

File: src/Enzo/enzo_EnzoConfig.cpp [hunk 5]
```diff
@@ -327,17 +339,47 @@ void EnzoConfig::pup (PUP::er &p)
...
+  if (method_grackle_use_grackle) {
+    p  | method_grackle_use_cooling_timestep;
+    p  | method_grackle_radiation_redshift;
 
-  if (method_grackle_use_grackle){
-    if (p.isUnpacking()) { method_grackle_chemistry = new chemistry_data; }
-    p | *method_grackle_chemistry;
-  } else {
-    method_grackle_chemistry = NULL;
+    //    int is_null = (method_grackle_chemistry==NULL);
```
Comment (aemerick) [inline src/Enzo/enzo_EnzoConfig.cpp:348]: Can remove this commented out code correct? (looks to me to bet the code from before the threadsafe Grackle PR

File: src/Enzo/enzo_EnzoConfig.cpp [hunk 6]
```diff
@@ -327,17 +339,47 @@ void EnzoConfig::pup (PUP::er &p)
...
+    //        }
+    //      }
+    //      ASSERT("EnzoConfig::pup",
+    //             "method_grackle_chemistry is expected to be non-null",
+    //             (method_grackle_chemistry != NULL));
+    //      
+    //      p | *method_grackle_chemistry;
+    //    }
+
+    if (method_grackle_use_grackle){
```
Comment (aemerick) [inline src/Enzo/enzo_EnzoConfig.cpp:376]: This if statement is redundant with the one above.

File: src/Cello/problem_MethodNull.cpp [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+#include "problem.hpp"
+
+void MethodNull::compute( Block * block) throw()
+{  block->compute_done(); }
+
+//======================================================================
+
+void MethodNull::init_refresh_()
+{
+  cello::simulation()->new_refresh_set_name(ir_post_,name());
```
[the remark this answers is not in the document]
Comment (jobordner) [reply src/Cello/problem_MethodNull.cpp:16]: The purpose of the Null method is in part to refresh fields at the beginning of the cycle immediately after the "adapt" phase.  And yes, refreshing all fields would make more sense.  Done.

File: src/Cello/control_new_refresh.cpp [hunk 8]
```diff
@@ -0,0 +1,816 @@
+// See LICENSE_CELLO file for license and copyright information
```
[the remark this answers is not in the document]
Comment (aemerick) [reply src/Cello/control_new_refresh.cpp:1]: Ah ok. That makes sense.

File: src/Cello/data_FieldData.cpp [hunk 9]
```diff
@@ -44,7 +44,7 @@ FieldData::FieldData
 FieldData::~FieldData() throw()
 {  
   deallocate_permanent();
-  for (int i=0; i<array_temporary_.size(); i++) {
+  for (size_t i=0; i<array_temporary_.size(); i++) {
     delete [] array_temporary_[i];
     array_temporary_[i] = NULL;
```
[the remark this answers is not in the document]
Comment (aemerick) [reply src/Cello/data_FieldData.cpp:49]: Thats fine with me!

File: src/Cello/control_refresh.cpp [hunk 10]
```diff
@@ -26,28 +26,8 @@
```
[the remark this answers is not in the document]
Comment (aemerick) [reply src/Cello/control_refresh.cpp:26]: Resolving this comment. Decided to move renaming to a new PR

File: src/Enzo/enzo_EnzoMethodPmDeposit.cpp [hunk 11]
```diff
@@ -97,221 +107,243 @@ void EnzoMethodPmDeposit::compute ( Block * block) throw()
...
+      int level = block->level();
 
-    // Accumulated single velocity array for Baryon deposit
-
-    for (int ib=0; ib<particle.num_batches(it); ib++) {
+      // Required for Cosmology ("mass" is mass)
+      // Not for Collapse ("mass" is density)
+      if (cosmology) {
+        dens *= std::pow(2.0,rank*level);
+      }
```
[the remark this answers is not in the document]
Comment (aemerick) [reply src/Enzo/enzo_EnzoMethodPmDeposit.cpp:162]: @jobordner I'm ok with waiting to address this. But yes, making an issue for it would be good. I can imagine this being a potential non-obvious pitfall for someone down the line.

File: input/Hi.in [hunk 12]
```diff
@@ -56,6 +56,17 @@
...
+   Group {
+     list = [ "conserved", "make_field_conservative" ];
+     conserved {
+         field_list = [ "density", "velocity_x", "velocity_y",
+                        "internal_energy", "total_energy"  ];
+     }
+     make_field_conservative {
+         field_list = [ "velocity_x", "velocity_y",
+                        "internal_energy", "total_energy" ];
+     }
```
Comment (pgrete) [inline input/Hi.in:68]: From a physical point of view, the total mass, momenta and total energy should be conserved.
Am I misinterpreting the naming (and could be make this potentially more clear) or should different fields be conserved?
Potentially related to #56

File: input/HelloWorld.in [hunk 13]
```diff
@@ -54,7 +54,19 @@
...
+     list = [ "conserved", "make_field_conservative" ];
+     conserved {
+         field_list = [ "density", "velocity_x", "velocity_y",
+                        "internal_energy", "total_energy" ];
+     }
+     make_field_conservative {
+         field_list = [ "velocity_x", "velocity_y",
+                        "internal_energy", "total_energy" ];
+     }
+  }
```
Comment (pgrete) [inline input/HelloWorld.in:69]: Same comment as in `Hi.in` regarding the conserved quantities.

File: input/checkpoint_ppm-1.in [hunk 14]
```diff
@@ -8,8 +8,7 @@ Mesh { root_blocks    = [1,1]; }
 include "input/adapt_slope.incl"
 
 Testing {
-    time_final = [0.00632976560208544,  # OLD_PPM
-                  0.00634160320623778]; # NEW_PPM
+    time_final = [0.00634502580849086];
```
Comment (pgrete) [inline input/checkpoint_ppm-1.in:11]: Is this value different from both OLD and NEW due to bug fixes?

File: input/output-stride.incl [hunk 15]
```diff
@@ -22,8 +22,7 @@ Output {
 Stopping {  cycle = 20; }
 Testing {   cycle_final = 20; }
 Testing {
-   time_final  = [0.0506399682687848, # OLD_PPM
-                  0.0507272430695593]; # NEW_PPM
+   time_final  = [0.0507272430695593];
```
[the remark this answers is not in the document]
Comment (aemerick) [reply input/output-stride.incl:25]: OLD_PPM referred to -- well -- old PPM methods that are no longer present in the code. This got removed in PR#24

[the remark this answers is not in the document]
Comment (aemerick) [reply input/output-stride.incl:25]: OLD_PPM referred to -- well -- old PPM methods that are no longer present in the code. This got removed in PR#24

[the remark this answers is not in the document]
Comment (pgrete) [reply input/output-stride.incl:25]: This comment referred to another comment above (in my interface hidden and shows only after "show" more) because here it seems that the results of NEW_PPM are the latest one.
At other places the "latest" does neither match NEW nor OLD.

File: SConstruct [hunk 16]
```diff
@@ -191,65 +185,59 @@ define = {}
...
 # Library defines
 
-define_hdf5  =        []
+define_png   =        'NO_FREETYPE'
 
-define_png   =        ['NO_FREETYPE']
+# Python version defines
 
-# Charm defines
+define_python_lt_27 = 'CONFIG_PYTHON_LT_27'
```
[the remark this answers is not in the document]
Comment (mabruzzo) [reply SConstruct:232]: The current installed version of SCons shipped with the codebase requires python 2.7. PR #46 is trying to address this issue

File: input/Hi.in (hunk 12 above)
Comment (jobordner) [reply input/Hi.in:68]: You're right, these fields are those whose fluxes are provided by hydro and are required in maintaining conservation, but not necessarily conserved themselves.  And, given that specifying field names in parameter files is being depreciated, these field groups should not be required at all.  I'll deal with the first issue for this PR, and defer the second.

File: src/Cello/control_new_refresh.cpp [hunk 17]
```diff
@@ -0,0 +1,932 @@
...
+
+    const int nb = particle.num_batches(it);
+
+    for (int ib=0; ib<nb; ib++) {
+
+      const int np = particle.num_particles(it,ib);
+
+      // ...extract particle position arrays
+
+      double xa[np],ya[np],za[np];
```
Comment (pgrete) [inline src/Cello/control_new_refresh.cpp:774]: `np` not known at compile time.
```suggestion
      std::vector<double> xa(np) ,ya(np) ,za(np) ;
```
Likely also requires update to the `position` call below.

File: src/Cello/error_Error.hpp [hunk 18]
```diff
@@ -245,6 +245,8 @@
   {  if (!(A)) { m2_(stderr,"ERROR",__FILE__,__LINE__,F,M,A1,A2,A3,A4,A5,A6); t_(); } }
 #define ASSERT7(F,M,A1,A2,A3,A4,A5,A6,A7,A)				\
   {  if (!(A)) { m2_(stderr,"ERROR",__FILE__,__LINE__,F,M,A1,A2,A3,A4,A5,A6,A7); t_(); } }
+#define ASSERT8(F,M,A1,A2,A3,A4,A5,A6,A7,A8,A)				\
+  {  if (!(A)) { m2_(stderr,"ERROR",__FILE__,__LINE__,F,M,A1,A2,A3,A4,A5,A6,A7,A8); t_(); } }
```
Comment (pgrete) [inline src/Cello/error_Error.hpp:249]: Would be great if the arguments are be protected by `()`, but likely sth for a later PR.

File: src/Enzo/enzo_EnzoMethodCheckGravity.cpp [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+
+	  int ix0 = gx + floor((nx-1)*(x - xm) / (xp - xm));
+	  int ix1 = ix0 + 1;
+	  double x0 = xm + (ix0-gx+0.5)*hx;
+	  double x1 = 1 - x0;
+	  double v0 = is_single ? vxa4[ix0] : vxa8[ix0];
+	  double v1 = is_single ? vxa4[ix1] : vxa8[ix1];
+	  double vx = v0*x1 + v1*x0;
+
+          // ...
```
Comment (pgrete) [inline src/Enzo/enzo_EnzoMethodCheckGravity.cpp:105]: Is this indicative of missing code?
Also applies to places further down in this file.

File: src/Enzo/enzo_EnzoMethodCheckGravity.cpp [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+
+    const double hx = (xp-xm)/nx;
+    const double hy = (yp-ym)/ny;
+    const double hz = (zp-zm)/nz;
+
+    double dt = block -> dt();
+
+    // Get velocity precision: ASSUMES precision same for all axes and is single or double
+    
+    const bool is_single = (field.precision (field.field_id("velocity_x")) == precision_single);
```
Comment (pgrete) [inline src/Enzo/enzo_EnzoMethodCheckGravity.cpp:75]: What happens if compiled with quadruple prec? Should there be a check here (or is there already somewhere else I missed) to make sure the code fails with an informative error message if quadruple is used?

**Review by forrestglines — COMMENTED** [review]
For what I've seen, looks good! The PR is a bit too large to understand everything, but we can fix problems as they arise later.

File: src/Cello/problem_Problem.cpp [hunk 21]
```diff
@@ -444,6 +446,8 @@ void Problem::initialize_method ( Config * config ) throw()
   const size_t num_method = config->method_list.size();
 
   Method::courant_global = config->method_courant_global;
+
+  method_list_.push_back(new MethodNull(config->method_null_dt));
```
Comment (mabruzzo) [inline src/Cello/problem_Problem.cpp:450]: This addition (coupled with the change in `MethodNull` to refresh all fields) creates a large problem for the VL+CT solver (in PR #9 ).

Presently, issues arise when refresh is scheduled to refresh all fields and there are registered temporary fields that have been deallocated. This is an issue for the VL+CT solver because the current implementation of its `compute` method presently allocates and deallocates a series of temporary fields during every evaluation (of non-leaf blocks). For a host of reasons, I have plans to move away from this approach, but I'm not sure I'm going to be able to get to that immediately.

Is there way we can work around this in the short-term? Maybe we can add an instance of `MethodNull` to `method_list_` after the for-loop if `method_list_` contains no entries? (If you could explain why you added this line, I might be able to make more constructive suggestions).

File: src/Enzo/enzo_EnzoSolverJacobi.cpp [hunk 22]
```diff
@@ -10,14 +10,16 @@
...
 // #define DEBUG_TRACE_CYCLE 0
 
 #ifdef DEBUG_TRACE
 #  define TRACE_JACOBI(BLOCK,SOLVER,METHOD)			\
   if (BLOCK->cycle() >= DEBUG_TRACE_CYCLE) {			\
-    CkPrintf ("%s:%d %s %s TRACE_JACOBI %s\n",			\
-	      __FILE__,__LINE__,BLOCK->name().c_str(),SOLVER->name().c_str(),METHOD); \
+    CkPrintf ("%s:%d %s %s TRACE_JACOBI active %d %s\n",			\
+	      __FILE__,__LINE__,BLOCK->name().c_str(),(SOLVER?SOLVER->name().c_str():"Unknown"), \
+              (SOLVER?SOLVER->is_finest(BLOCK):-1),METHOD);             \
```
Comment (pgrete) [inline src/Enzo/enzo_EnzoSolverJacobi.cpp:22]:
```suggestion
	      __FILE__,__LINE__,BLOCK->name().c_str(),((SOLVER)?(SOLVER)->name().c_str():"Unknown"), \
              ((SOLVER)?(SOLVER)->is_finest((BLOCK)):-1),(METHOD));             \
```

File: src/Enzo/enzo_SolveHydroEquations.cpp [hunk 23]
```diff
@@ -8,82 +8,19 @@
...
-      }								\
-    }								\
-    fclose(fp);							\
-    fp=NULL;							\
-  }
+// #define IE_ERROR_FIELD
+// #define DEBUG_PPM
+
+#ifdef DEBUG_PPM
+#   define TRACE(MSG,BLOCK) CkPrintf ("DEBUG_PPM: %s %s\n",BLOCK->name().c_str(),MSG); fflush(stdout);
```
Comment (pgrete) [inline src/Enzo/enzo_SolveHydroEquations.cpp:18]:
```suggestion
#   define TRACE(MSG,BLOCK) CkPrintf ("DEBUG_PPM: %s %s\n",(BLOCK)->name().c_str(),(MSG)); fflush(stdout);
```

File: src/Enzo/fortran_types.h [hunk 24]
```diff
@@ -2,8 +2,8 @@
 #ifdef FORTRAN
 #  ifdef CONFIG_PRECISION_SINGLE
 #     define CONFIG_BFLOAT_4
-#     define tiny 1.d-20
-#     define huge 1.d+20
+#     define tiny 1.e-20
+#     define huge 1.e+20
```
Comment (pgrete) [inline src/Enzo/fortran_types.h:6]: Why the change only for single prec but not double or quad (asking as a non-Fortran person)?

**Review by pgrete — CHANGES_REQUESTED** [review]
I now finished reviewing all files.
This is a tremendous amount of work!
Please excuse all the nitpicking comments I left.
I know how important it is to get this into master.
I'm only requesting changes as I think the variable array initialization may actually be/lead to a bug (please confirm).
Feel free to ignore all the comments on "code modernization" (which I think are still important but can probably addressed outside of this PR).

File: src/Cello/problem_Problem.cpp (hunk 21 above)
Comment (jobordner) [reply src/Cello/problem_Problem.cpp:450]: How is the problem manifesting itself? The purpose of inserting MethodNull at the beginning of the list of methods is (in part) to ensure all fields are refreshed after the adapt phase.  It should only affect leaf blocks, I don't know why you're having problems with non-leaf blocks. Moving the refresh step to the end of the adapt phase would be cleaner, but functionally identical, so that wouldn't solve anything. Would skipping temporary fields when using add_all_fields() solve the conflict?

Comment (mabruzzo) [reply src/Cello/problem_Problem.cpp:450]: Apologies for the delayed response.

The problem manifests during refresh in two different ways based on the centering of the first temporary field that the refresh machinery handles:
1. If the first handled temporary field has `cx`,`cy`,`cz` values that are all 0 or 1, a segmentation fault occurs (because the refresh machinery is trying to access a `NULL` pointer). The traceback says this occurs in `FieldFace::face_to_face`
2. If any of the `cx`,`cy`,`cz` values are `-1`, than it triggers an error (in `FieldFace::face_to_face`'s call to `FieldFace::loop_limits`) that I previously added to the master branch that informs the user that refreshing this type of face-centered field is not currently supported.

I don't think that the leaf/non-leaf blocks have anything to do with this problem since I have not currently not tested AMR. I think I may have accidentally obfuscated the issue by mentioning this at all.

I think skipping temporary fields when using `add_all_fields()` would solve the problem. Would this actually involve changing `FieldFace::field_list_src_` and `FieldFace::field_list_dest_`.

If you want to replicate this problem on this branch, just replace the implementation of `MethodNull::compute` with:

```c++
void MethodNull::compute( Block * block) throw(){
  std::vector<int> id_l;
  std::vector<std::string> names = {"temp0", "temp1", "temp2"};
  std::vector<int> cz_l = {0,-1,1};
  Field field = block->data()->field();
  FieldDescr * field_descr = field.field_descr();

  // first reserve and allocate temporary fields:
  if (field.is_field(names[0])){
    // the temporary field was added in a previous time-step
    for (std::string name : names){ id_l.push_back(field.field_id(name)); }
  } else {
    // Reserve a temporary field
    for (std::size_t i=0; i<names.size(); i++){
      id_l.push_back(field_descr->insert_temporary(names[i]));
      field_descr->set_centering(id_l[i],0,0,cz_l[i]);
    }
  }
  for (int id : id_l){ field.allocate_temporary(id); }

  // --- do some work ---

  // deallocate the temporary field:
  for(int id : id_l){ field.deallocate_temporary(id); }

  block->compute_done();
}
```

Comment (mabruzzo) [reply src/Cello/problem_Problem.cpp:450]: Skipping temporary fields when using `add_all_fields()` appears to solve the problem. I had modified `FieldFace::field_list_src_` and `FieldFace::field_list_dest_` to only add fields (in the for-loop) if they are permanent. My VL+CT tests now pass.

I'm now verifying that this doesn't cause the rest of the test-suite to break

**Review by pgrete — APPROVED** [review]
Thanks for addressing the comments.
Pending a fix of the potential typo  (to be confirmed), I'm happy to approve.
Also, in the future, it'd be great if we make more use of the discussion/comment feature in PRs, which would make (re-)reviewing a lot easier.
Specifically, I'm thinking about replying to all comments left by reviewers so that it's easy to see whether the comment has been addressed or not (and why/how).

**Review by bwoshea — APPROVED** [review]

**Comment by bwoshea** [comment]
3 approves, I'm merging.  WOOHOO!
